DylibBinder/DBGenericParameter.cs

```diff
@@ -0,0 +1,153 @@
+		}
+	}
+
+	internal class DBGenericParameters {
+
+		public DBGenericParameters (SwiftBaseFunctionType signature) {
+			AssignGenericParameters (signature);
+		}
+
+		public DBGenericParameters (SwiftType swiftType)
+		{
+			FilterGenericParameterSource (swiftType);
+		}
+
+		public DBGenericParameters (DBTypeDeclaration typeDeclaration)
+		{
+			ParseTopLevelGenerics (typeDeclaration.Funcs, typeDeclaration.Properties);
+		}
+
+		public List<DBGenericParameter> GenericParameters { get; } = new List<DBGenericParameter> ();
+
+
+		void AssignGenericParameters (SwiftBaseFunctionType signature)
+		{
+			FilterGenericParameterSource (signature.Parameters, signature.ReturnType);
+		}
+
+		void FilterGenericParameterSource (params SwiftType [] types)
+		{
+			foreach (var type in types) {
+				switch (type) {
+				case SwiftFunctionType funcType:
+					HandleType (funcType);
+					return;
+				case SwiftBoundGenericType boundType:
+					HandleType (boundType);
+					return;
+				case SwiftTupleType tupleType:
+					HandleType (tupleType);
+					return;
+				case SwiftGenericArgReferenceType refType:
+					HandleType (refType);
+					return;
+				default:
+					return;
+				}
+			}
+		}
+
+		void HandleType (SwiftFunctionType funcType)
+		{
+			switch (funcType.Parameters) {
+			case SwiftBoundGenericType boundType:
+				HandleType (boundType);
+				return;
+			case SwiftTupleType tupleType:
+				HandleType (tupleType);
+				return;
+			case SwiftGenericArgReferenceType refType:
+				HandleType (refType);
+				return;
+			default:
+				return;
+			}
+		}
+
+		void HandleType (SwiftBoundGenericType boundType)
+		{
+			foreach (var bound in boundType.BoundTypes) {
+				if (bound is SwiftGenericArgReferenceType refType)
+					HandleType (refType);
+			}
+		}
+
+		void HandleType (SwiftTupleType tupleType)
+		{
+			foreach (var content in tupleType.Contents) {
+				switch (content) {
+				case SwiftBoundGenericType boundGType:
+					HandleType (boundGType);
+					return;
+				case SwiftFunctionType cFuncType:
+					HandleType (cFuncType);
+					return;
+				case SwiftGenericArgReferenceType refType:
+					HandleType (refType);
+					return;
+				case SwiftTupleType innerTupleType:
+					HandleType (innerTupleType);
+					return;
+				default:
+					return;
+				}
+			}
+		}
+
+		void HandleType (SwiftGenericArgReferenceType refType)
+		{
+			AddIfNotPresent (new DBGenericParameter (refType.Depth, refType.Index));
+		}
+
+		void AddIfNotPresent (DBGenericParameter newGP)
+		{
+			if (GenericParameters.Count == 0) {
+				GenericParameters.Add (newGP);
+				return;
+			}
+
+			if (!GenericParameters.Contains (newGP))
+				GenericParameters.Add (newGP);
+		}
+
+		void ParseTopLevelGenerics (DBFuncs funcs, DBProperties properties)
+		{
+			foreach (var func in funcs.Funcs) {
+				GrabTopLevelGenerics (func.GenericParameters);
+			}
+
+			foreach (var prop in properties.Properties) {
+				GrabTopLevelGenerics (prop.GenericParameters);
+			}
```

Comment:
Are funcs and properties the same type or a parent type? We could use Concat and have a single loop. Not that it matters, I'm having fun :)

Reply:
Ah no they have different types/parent types